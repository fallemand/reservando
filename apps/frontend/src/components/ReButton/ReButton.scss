@import "~@/assets/styles/variables";
@import "./buttons";

.re-button {
  font-weight: $font-weight-bold;
  cursor: pointer;
  font-family: inherit;
  padding: 1em;
  text-decoration: none;

  &-xsmall {
    font-size: 0.5em;
  }

  &-small {
    font-size: 14px;
    padding: $bdu;
  }

  &-medium {
    font-size: 1em;
  }

  &-large {
    font-size: 1.5em;
  }

  &-with-icon {
    display: inline-flex;
    align-items: center;
    justify-content: center;
  }

  &-icon {
    display: flex;
    margin-right: 0.5em;

    .svg-icon {
      display: flex;
    }
  }
}

.re-button-cta {
  &-red {
    @include cta-generator($red-link, darken($red-link, 10%), black);
  }
  &-uva {
    @include cta-generator($grape, darken($grape, 10%), black);
  }
  // &-success {
  //   @include cta-generator($success, $success-active, $midnight-blue);
  // }
  // &-error {
  //   @include cta-generator($danger, $danger-active, $midnight-blue);
  // }
  // &-white {
  //   @include cta-generator($white, darken($white, 10%), $ui-blue, $ui-blue);
  // }
}

.re-button-outline {
  &-uva {
    @include outline-generator($grape, darken($grape, 10%), black);
  }
}
//   &-dark {
//     @include outline-generator($midnight-blue, darken($midnight-blue, 10%), $ui-blue);
//   }
//   &-success {
//     @include outline-generator($success, $success-active, $midnight-blue);
//   }
//   &-error {
//     @include outline-generator($danger, $danger-active, $midnight-blue);
//   }
//   &-white {
//     @include outline-generator($white, darken($white, 10%), $ui-blue);
//   }
// }

// .re-button-link {
//   &-light {
//     @include link-generator($cta, $cta-active, $midnight-blue);
//   }
//   &-dark {
//     @include link-generator($midnight-blue, darken($midnight-blue, 10%), $ui-blue);
//   }
//   &-success {
//     @include link-generator($success, $success-active, $midnight-blue);
//   }
//   &-error {
//     @include link-generator($danger, $danger-active, $midnight-blue);
//   }
//   &-white {
//     @include link-generator($white, darken($white, 10%), $ui-blue);
//   }
// }
