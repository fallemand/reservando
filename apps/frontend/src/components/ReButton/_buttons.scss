// -------------------------------------------------------------------------------
//  Buttons
//
// üíÅ‚Äç use variables rather than hard-coded pixel/REM values
// -------------------------------------------------------------------------------

$btn-border-radius: 5rem;
$btn-border-width: 2px;
$btn-border-style: solid;

/**
* Cta styles
*/
@mixin cta-generator($color, $hover-color, $focus-color, $font-color: $white) {
  color: $font-color;
  background: $color;
  border-color: $color;
  border-style: $btn-border-style;
  border-radius: $btn-border-radius;
  border-width: $btn-border-width;
  transition: background-color 0.2s ease-out;
  outline: 0;

  &:hover,
  &:active {
    color: $font-color;
    background: $hover-color;
    border-color: $hover-color;
  }

  &:disabled {
    color: $white;
    background: $border-color;
    border-color: $border-color;
    cursor: not-allowed;
  }

  &:focus:not(:hover) {
    color: $font-color;
    background: $color;
    border-color: $focus-color;
  }
}

/**
* Outline styles
*/
@mixin outline-generator($color, $hover-color, $focus-color) {
  color: $color;
  background: transparent;
  border-color: $color;
  border-style: $btn-border-style;
  border-radius: $btn-border-radius;
  border-width: $btn-border-width;
  transition: border-color 0.2s ease-out;
  outline: 0;

  &:hover,
  &:active {
    color: $hover-color;
    background: transparent;
    border-color: $hover-color;
  }

  &:disabled {
    color: $border-color;
    border-color: $border-color;
    background: transparent;
    cursor: not-allowed;
  }

  &:focus:not(:hover) {
    border-color: $focus-color;
  }
}

/**
* Link styles
*/
@mixin link-generator($color, $hover-color, $focus-color) {
  text-decoration: none;
  background: transparent;
  border: 0;
  color: $color;
  transition: color 0.2s ease-out;
  padding: 0.25em 0.5em;
  outline: 0;

  &:hover,
  &:active {
    color: $hover-color;
  }

  &:disabled {
    color: $border-color;
    background: transparent;
    cursor: not-allowed;
  }

  &:focus:not(:hover) {
    color: $focus-color;
  }
}
