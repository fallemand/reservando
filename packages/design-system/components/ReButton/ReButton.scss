@import "../../styles/variables.scss";
@import "./buttons.scss";

.re-button {
  font-weight: $font-weight-bold;
  cursor: pointer;
  font-family: inherit;
  padding: 1em;
  text-decoration: none;

  &--small {
    font-size: 14px;
    padding: $bdu*1 $bdu*1.5;
  }

  &--medium {
    font-size: 1em;
    padding: $bdu*1.5 $bdu*2.5;
  }

  &--large {
    font-size: 1em;
    padding: $bdu*2 $bdu*3;
  }

  &--with-icon {
    display: inline-flex;
    align-items: center;
    justify-content: center;
  }

  &-icon {
    display: flex;
    margin-right: 0.5em;

    .svg-icon {
      display: flex;
    }
  }

  &--primary {
    @include cta-generator($primary-500, $primary-600,$primary-700, black, $white, $grey-300);
  }

  &--secondary {
    @include cta-generator($primary-100, $primary-200, $primary-300, black, $primary-500, transparent);
  }

  &--secondary-outline {
    @include cta-generator(transparent, $primary-100, $primary-200, black, $primary-500, transparent, $primary-500);
  }

  &--link {
    @include link-generator($primary-500, $primary-600, $primary-700, black);
  }
}
