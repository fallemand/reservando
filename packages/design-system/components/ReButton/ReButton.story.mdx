import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import { ReButton } from "@reservando/design-system";

<Meta
  title="Components/Button"
  component={ReButton}
  argTypes={{
    size: { control: { type: "select", options: ["small", "medium", "large"] } },
    modifier: { control: { type: "select", options: ["cta-red", "cta-uva", "outline-uva"] } },
    type: { type: "select", options: ["button", "submit"] },
  }}
/>

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { ReButton },
  template: '<ReButton :size="size" :modifier="modifier">{{label}}</ReButton>',
});

# ReButton

<Canvas>
  <Story name="basic">
    {{
      components: { ReButton },
      template: "<ReButton>My Button</ReButton>",
    }}
  </Story>
</Canvas>

<ArgsTable of={ReButton} />

## How to use

```html
<template>
  <ReButton @click="clickHandler">Label</ReButton>
</template>

<script lang="ts">
  import { ReButton } from "@reservando/design-system";
</script>
```

## Using modifiers

<Canvas>
  <Story name="modifiers">
    {{
      components: { ReButton },
      data() {
        return {
          checked: true,
        };
      },
      template: `
      <div>
        <ReButton modifier="primary" class="mr-2">primary</ReButton>
        <ReButton modifier="secondary" class="mr-2">secondary</ReButton>
        <ReButton modifier="secondary-outline" class="mr-2">secondary-outline</ReButton>
        <ReButton modifier="link" class="mr-2">link</ReButton>
        <p class="hint mt-4">Disabled</p>
        <ReButton modifier="primary" disabled="true" class="mr-2">primary</ReButton>
        <ReButton modifier="secondary" disabled="true" class="mr-2">secondary</ReButton>
        <ReButton modifier="secondary-outline" disabled="true" class="mr-2">secondary-outline</ReButton>
        <ReButton modifier="link" disabled="true" class="mr-2">link</ReButton>
      </div>
    `,
    }}
  </Story>
</Canvas>

## Using sizes

<Canvas>
  <Story name="sizes">
    {{
      components: { ReButton },
      data() {
        return {
          checked: true,
        };
      },
      template: `
      <div>
        <ReButton size="small" class="mr-2">small</ReButton>
        <ReButton size="medium" class="mr-2">medium</ReButton>
        <ReButton size="large" class="mr-2">large</ReButton>
      </div>
    `,
    }}
  </Story>
</Canvas>
