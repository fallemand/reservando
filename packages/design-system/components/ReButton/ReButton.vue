<template>
  <component
    :is="renderAsLink($attrs) ? 'a' : 'button'"
    :class="[
      're-button',
      `re-button-${size}`,
      `re-button-${modifier}`,
      { 're-button-with-icon': $slots.icon },
    ]"
    v-bind="$attrs"
    :type="renderAsLink($attrs) ? null : type"
  >
    <span v-if="$slots.icon" class="re-button-icon">
      <!-- @slot Icon at the beginning of the button -->
      <slot name="icon" />
    </span>

    <slot />
  </component>
</template>

<script lang="ts">
import { defineComponent, PropType } from "vue";

export const TYPES = ["button", "submit", "reset"];
export const SIZES = ["xsmall", "small", "medium", "large"];
export const MODIFIERS = ["cta-uva", "cta-red", "outline-uva"];

const ReButton = defineComponent({
  name: "ReButton",
  props: {
    size: {
      default: "medium",
      type: String as PropType<typeof SIZES[number]>,
      validator: (value: string): boolean => SIZES.includes(value),
    },
    modifier: {
      default: "cta-red",
      type: String as PropType<typeof MODIFIERS[number]>,
      validator: (value: string): boolean => MODIFIERS.includes(value),
    },
    type: {
      default: "button",
      type: String as PropType<typeof TYPES[number]>,
      validator: (value: string): boolean => TYPES.includes(value),
    },
  },
  methods: {
    renderAsLink(attrs?: { href: string }): boolean {
      return !!attrs && !!attrs.href;
    },
  },
});

export default ReButton;
</script>

<style lang="scss" src="./ReButton.scss" />
