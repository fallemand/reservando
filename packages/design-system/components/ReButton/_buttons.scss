// -------------------------------------------------------------------------------
//  Buttons
//
// üíÅ‚Äç use variables rather than hard-coded pixel/REM values
// -------------------------------------------------------------------------------

$btn-border-width: 1px;
$btn-border-style: solid;

/**
* Cta styles
*/
@mixin cta-generator($color, $hover-color, $focus-color, $font-color: $white) {
  color: $font-color;
  background: $color;
  border-color: $color;
  border-style: $btn-border-style;
  border-radius: 5rem;
  border-width: $btn-border-width;
  transition: background-primary 0.2s ease-out;
  outline: 0;

  &:hover,
  &:active {
    color: $font-color;
    background: $hover-color;
    border-color: $hover-color;
  }

  &:disabled {
    color: $white;
    background: $border-secondary;
    border-color: $border-secondary;
    cursor: not-allowed;
  }

  &:focus:not(:hover) {
    color: $font-color;
    background: $color;
    border-color: $focus-color;
  }
}

/**
* Outline styles
*/
@mixin outline-generator($color, $hover-color, $focus-color) {
  color: $color;
  background: transparent;
  border-color: $color;
  border-style: $btn-border-style;
  border-radius: $bdu * 1.5;
  border-width: $btn-border-width;
  transition: border-color 0.2s ease-out;
  outline: 0;

  &:hover,
  &:active {
    color: $hover-color;
    background: transparent;
    border-color: $hover-color;
  }

  &:disabled {
    color: $border-secondary;
    border-color: $border-secondary;
    background: transparent;
    cursor: not-allowed;
  }

  &:focus:not(:hover) {
    border-color: $focus-color;
  }
}

/**
* Link styles
*/
@mixin link-generator($color, $hover-color, $focus-color) {
  text-decoration: none;
  background: transparent;
  border: 0;
  color: $color;
  transition: color 0.2s ease-out;
  padding: 0.25em 0.5em;
  outline: 0;

  &:hover,
  &:active {
    color: $hover-color;
  }

  &:disabled {
    color: $border-secondary;
    background: transparent;
    cursor: not-allowed;
  }

  &:focus:not(:hover) {
    color: $focus-color;
  }
}
