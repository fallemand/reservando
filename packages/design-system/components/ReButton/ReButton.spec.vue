<script lang="ts">
import { shallowMount } from "@vue/test-utils";
import ReButton from "./ReButton.vue";

describe("ReButton component", () => {
  it("should render with default props", () => {
    const wrapper = shallowMount(ReButton, {
      slots: {
        default: "__BUTTON-TEXT__",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render disabled button", () => {
    const wrapper = shallowMount(ReButton, {
      attrs: {
        disabled: true,
      },
      slots: {
        default: "__BUTTON-TEXT__",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render with props", () => {
    const wrapper = shallowMount(ReButton, {
      props: {
        modifier: "secondary",
        size: "large",
        type: "submit",
      },
      slots: {
        default: "__BUTTON-TEXT__",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render as <a /> when href is passed", () => {
    const wrapper = shallowMount(ReButton, {
      attrs: {
        href: "http://fake-url.com",
        target: "_blank",
      },
      slots: {
        default: "__BUTTON-TEXT__",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render with icon", () => {
    const wrapper = shallowMount(ReButton, {
      slots: {
        default: "__BUTTON-TEXT__",
        icon: "__ICON__",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should bind all listeners to the button", () => {
    const listeners = {
      onClick: jest.fn(),
      onMouseleave: jest.fn(),
      onMouseover: jest.fn(),
    };
    const wrapper = shallowMount(ReButton, {
      attrs: listeners,
    });
    const input = wrapper.find("button");
    input.trigger("click");
    input.trigger("mouseleave");
    input.trigger("mouseover");
    expect(listeners.onClick).toHaveBeenCalled();
    expect(listeners.onMouseleave).toHaveBeenCalled();
    expect(listeners.onMouseover).toHaveBeenCalled();
  });
});
</script>
