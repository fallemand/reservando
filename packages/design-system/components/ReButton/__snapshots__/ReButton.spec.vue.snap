// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReButton component should render as <a /> when href is passed 1`] = `
<a
  class="re-button re-button--medium re-button--primary"
  href="http://fake-url.com"
  target="_blank"
>
  <!--v-if-->
  
  __BUTTON-TEXT__
  
</a>
`;

exports[`ReButton component should render disabled button 1`] = `
<button
  class="re-button re-button--medium re-button--primary"
  disabled=""
  type="button"
>
  <!--v-if-->
  
  __BUTTON-TEXT__
  
</button>
`;

exports[`ReButton component should render with default props 1`] = `
<button
  class="re-button re-button--medium re-button--primary"
  type="button"
>
  <!--v-if-->
  
  __BUTTON-TEXT__
  
</button>
`;

exports[`ReButton component should render with icon 1`] = `
<button
  class="re-button re-button--medium re-button--primary re-button--with-icon"
  type="button"
>
  <span
    class="re-button-icon"
  >
    <!-- @slot ReIcon at the beginning of the button -->
    
    __ICON__
    
  </span>
  
  __BUTTON-TEXT__
  
</button>
`;

exports[`ReButton component should render with props 1`] = `
<button
  class="re-button re-button--large re-button--secondary"
  type="submit"
>
  <!--v-if-->
  
  __BUTTON-TEXT__
  
</button>
`;
